File: sample/java/08_langtons_ant.java
final class _08_langtons_ant {
    private _08_langtons_ant() {
    }


    // 08: Sample that outputs Langton's Ant trajectories as a GIF.

    public static java.util.ArrayList<Long> capture(java.util.ArrayList<java.util.ArrayList<Long>> grid, long w, long h) {
        java.util.ArrayList<Long> frame = PyRuntime.__pytra_bytearray(w * h);
        for (long y = 0L; y < h; y += 1L) {
            long row_base = y * w;
            for (long x = 0L; x < w; x += 1L) {
                frame.set((int)((((row_base + x) < 0L) ? (((long)(frame.size())) + (row_base + x)) : (row_base + x))), (((((Long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).get((int)((((x) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).size())) + (x)) : (x)))))) != 0L)) ? (255L) : (0L)));
            }
        }
        return PyRuntime.__pytra_bytearray(frame);
    }

    public static void run_08_langtons_ant() {
        long w = 420L;
        long h = 420L;
        String out_path = "sample/out/08_langtons_ant.gif";
        double start = time.perf_counter();
        java.util.ArrayList<java.util.ArrayList<Long>> grid = new java.util.ArrayList<java.util.ArrayList<Long>>();
        for (long __ = 0L; __ < h; __ += 1L) {
            grid.add(PyRuntime.__pytra_list_repeat(0L, w));
        }
        long x = w / 2L;
        long y = h / 2L;
        long d = 0L;
        long steps_total = 600000L;
        long capture_every = 3000L;
        java.util.ArrayList<java.util.ArrayList<Long>> frames = new java.util.ArrayList<java.util.ArrayList<Long>>();
        for (long i = 0L; i < steps_total; i += 1L) {
            if (((((Long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).get((int)((((x) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).size())) + (x)) : (x))))))) == (0L))) {
                d = (d + 1L) % 4L;
                ((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).set((int)((((x) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).size())) + (x)) : (x))), 1L);
            } else {
                d = (d + 3L) % 4L;
                ((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).set((int)((((x) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).size())) + (x)) : (x))), 0L);
            }
            if (((d) == (0L))) {
                y = (y - 1L + h) % h;
            } else {
                if (((d) == (1L))) {
                    x = (x + 1L) % w;
                } else {
                    if (((d) == (2L))) {
                        y = (y + 1L) % h;
                    } else {
                        x = (x - 1L + w) % w;
                    }
                }
            }
            if (((i % capture_every) == (0L))) {
                frames.add(capture(grid, w, h));
            }
        }
        gif.save_gif(out_path, w, h, frames, gif.grayscale_palette(), 5L, 0L);
        double elapsed = time.perf_counter() - start;
        System.out.println(String.valueOf("output:") + " " + String.valueOf(out_path));
        System.out.println(String.valueOf("frames:") + " " + String.valueOf(((long)(frames.size()))));
        System.out.println(String.valueOf("elapsed_sec:") + " " + String.valueOf(elapsed));
    }
}

File: sample/julia/08_langtons_ant.jl
include(joinpath(@__DIR__, "built_in", "py_runtime.jl"))

include(joinpath(@__DIR__, "std", "time.jl"))
include(joinpath(@__DIR__, "utils", "gif.jl"))

# 08: Sample that outputs Langton's Ant trajectories as a GIF.

function capture(grid, w, h)
    frame = __pytra_bytearray((w * h))
    for y in 0:h - 1
        row_base = (y * w)
        for x in 0:w - 1
            frame[__pytra_idx((row_base + x), length(frame))] = (__pytra_truthy(grid[__pytra_idx(y, length(grid))][__pytra_idx(x, length(grid[__pytra_idx(y, length(grid))]))]) ? (255) : (0))
        end
    end
    return __pytra_bytes(frame)
end

function run_08_langtons_ant()
    w = 420
    h = 420
    out_path = "sample/out/08_langtons_ant.gif"
    
    start = perf_counter()
    grid = Any[]
    for _ in 0:h - 1
        push!(grid, __pytra_repeat_seq([0], w))
    end
    x = (w ÷ 2)
    y = (h ÷ 2)
    d = 0
    
    steps_total = 600000
    capture_every = 3000
    frames = Any[]
    
    for i in 0:steps_total - 1
        if (grid[__pytra_idx(y, length(grid))][__pytra_idx(x, length(grid[__pytra_idx(y, length(grid))]))] == 0)
            d = ((d + 1) % 4)
            grid[__pytra_idx(y, length(grid))][__pytra_idx(x, length(grid[__pytra_idx(y, length(grid))]))] = 1
        else
            d = ((d + 3) % 4)
            grid[__pytra_idx(y, length(grid))][__pytra_idx(x, length(grid[__pytra_idx(y, length(grid))]))] = 0
        end
        if (d == 0)
            y = (((y - 1) + h) % h)
        elseif (d == 1)
            x = ((x + 1) % w)
        else
            if (d == 2)
                y = ((y + 1) % h)
            else
                x = (((x - 1) + w) % w)
            end
        end
        if ((i % capture_every) == 0)
            push!(frames, capture(grid, w, h))
        end
    end
    save_gif(out_path, w, h, frames, grayscale_palette(), 5, 0)
    elapsed = (perf_counter() - start)
    __pytra_print("output:", out_path)
    __pytra_print("frames:", length(frames))
    __pytra_print("elapsed_sec:", elapsed)
end


run_08_langtons_ant();
